File: Ruby Programs/ALI/ali.rb
class ALI
  ProgramMemory = 128 #declares program memory and other memories based on requirements
  DataMemory = 128
  Memory = 256
  attr_accessor :memory, :accumulator, :DataRegister, :ProgramCounter, :ZeroBit, :instructionHash, :halted #gets getters for our data members

  def initialize#intiliazes our data members
    @memory = Array.new(Memory, 0) #creates a new array with the memory wanted
    @DataRegister = 0
    @accumulator = 0 #sets data all to 0
    @ProgramCounter = 0
    @opcode = ''
    @ZeroBit = false #starts with false on zero bit
    @instructionHash = {} #stores all of the variables declared
    @halted = false #sets halted to false
    @counter = 0 #to check if it does more than 1000 instructions
  end #end of initialize
  
  def allocateData #function to allocate memory for objects
    (DataMemory...Memory).each do |i| #traverses through memory array and checks if the current position is occupied
      return i unless @instructionHash.value?(i) #tries to find an unoccupied position
    end
    "Error: No available data memory" #else no space is available
  end

  def ReadFile(filename) #function to read in the instructions assuming its in correct format
    File.open(filename, 'r') do |file| #opens file to read
      file.each_with_index do |line, index|
        break if index >= ProgramMemory #stops reading in if it exceeds our program memory
        @memory[index] = line.strip #stores line into memory
      end
    end
  end #end of ReadFile

  def userLoop #takes user loop to either step or do all instructions at once
    command = ''
    puts "Welcome to my ALI interpreter:"
    while command != 'q' #keeps looping until user exits
      puts " Enter a command:
      a - Run all instructions at once:
      s - Step into next instruction
      q - Exit"
      command = gets.chomp #takes user input
      if command == 's' #checks if user wants to step
        executeLine #only runs one line
      elsif command == 'a' #checks if user wants to do all instructions
        executeAll #runs all instructions
      elsif command == 'q' #checks if user wants to quit
        puts "Exiting..."
        break
      else
        puts "INVALID INPUT. Enter only 'a','s', or 'q'"
      end
    end
  end #end of userLoop


  def executeLine #function that handles our 's' command
    return puts "Program has halted." if @halted #checks if program is halted before we try to execute line
    instruction = @memory[@ProgramCounter] #gets current instruction
    puts "Executing instruction at line=#{@ProgramCounter}:"
    executeInstruction(instruction) #runs only one instruction
    @ProgramCounter += 1 #adds 1 to counter
    @counter+=1
    printData #prints info of current memory
  end#end of executeLine

  def executeAll #function that handles 'a' command
    while @ProgramCounter < ProgramMemory && !@halted #makes sure we dont pass our program memory and that program has not halted
      @counter+=1
      if @counter >= 1000
        puts "want to continue enter y or n?"
        input = gets.chomp
        if input == 'n'
          @halted = true

        else input == 'y'
           @counter = 0
        end
      end
      instruction = @memory[@ProgramCounter]#gets instruction at current index
      executeInstruction(instruction)#executes the instruction
      @ProgramCounter += 1 #adds to counter to go to next instruction
    end
    printData #prints info of current memory
  end #end of executeAll

  def executeInstruction(instruction) #function to handle our instructions
    parts = instruction.split #splits the line between instruction and variable/number
    command = parts[0] #instruction will always be first
    if command == 'DEC' #checks if instruction is DEC
      variable = parts[1] #takes the variable name
      address = allocateData #creates a space in memory
      @instructionHash[variable] = address #assigns variable to the next address
      puts "Executed opcode: DEC" #printsOpcode only for this and LDI


    elsif command == 'LDA' #checks if command is LDA
      LDA.new(self, parts[1]).execute  #calls load command and passes the variable name of object wanted to load

    elsif command == 'LDI' #checks if command is LDI
      @accumulator = parts[1].to_i #loads integer into accumulator
      puts "Executed opcode: LDI" #prints opcode


    elsif command == 'STR' #checks if command is STR
      STR.new(self).execute(parts[1]) #calls store command and passes symbol name of what we want to store

    elsif command == 'ADD'
      ADD.new(self).execute  # No additional parameters needed

    elsif command == 'SUB'
      SUB.new(self, parts[1]).execute  # Pass the symbol to SubCommand

    elsif command == 'XCH' #checks if command is XCH
      return XCH.new(self).execute #calls X

    elsif command == 'JMP' #check if command is JMP
      JMP.new(self, parts[1].to_i).execute #calls jump command and passes the line number we want to jump too

    elsif command == 'JZS' #check if command is JZS
      JZS.new(self, parts[1].to_i).execute #calls jump if zero command and passes line number to jump too

    elsif command == 'HLT' #check if command is HLT
      HLT.new(self).execute #calls halt command

    else #anything else as instruction will lead to error
      puts "ERROR: #{command}"
    end
  end #end of executeInstruction

  def printData #function to prints all our important data
    puts "Accumulator: #{@accumulator}"
    puts "Data Register: #{@DataRegister}"
    puts "Program Counter: #{@ProgramCounter}"
    puts "Zero Bit: #{@ZeroBit}"
    puts "Memory: #{@memory[0...DataMemory]}"
    puts "Data Memory: #{@memory[DataMemory...Memory]}"
  end #end of printData

end #end of class ALI

File: C++ Programs/ALI/ALI.cpp
#include "ALI.h"
#include "Command.h"
#include <fstream>
#include <sstream>
#include <iostream>
using namespace std;

ALI::ALI(){//default constructor
    accumulator = 0; //sets all values to 0/false 
    dataRegister = 0;
    programCounter = 0; 
    zeroBit = false;
    halted = false;
    counter = 0;
    initialize();//fills in array with 0's
}

void ALI::initialize(){//definition of initialize
    fill(begin(programMemory), end(programMemory), 0);//fill program memory will all 0's
    fill(begin(dataMemory), end(dataMemory), 0);//fill in data memory to zero
}

int ALI::allocateData(){
    static int nextAddress = 0;//keeps track of our addresses
    if (nextAddress < DATA_MEMORY_SIZE) {
        return nextAddress++;//returns next address
    } 
    else{//out of program memory so we exit
        exit(0);
    }
}

void ALI::loadProgram(const string& filename){//loadProgram definiton that reads in file 
    ifstream file(filename);
    string line;
    int index = 0;
    while (getline(file, line) && index < PROGRAM_MEMORY_SIZE) {
        programMemory[index] = line;//stores instruction into array memory
        index++;//making sure we dont go over 250
    }
    file.close();//closes file
}

void ALI::userLoop(){//userLoop definiton that keeps loop running until user leaves
    string command;
    cout << "Welcome to my ALI interpreter:" << endl;
    while (command != "q"){//loop keeps running until command is q
        cout << "Enter a command:\n a - Run all instructions at once\n s - Step into next instruction\n q - Exit" << endl;
        cin >> command;
        if (command == "s"){//check if its s
            executeLine();
        } 
        else if (command == "a"){//checks if its a
            executeAll();
        } 
        else if (command == "q") {//when user enters q then it breaks
            cout << "Exiting..." << endl;
            break;
        } 
        else{//else give error message
            cout << "INVALID INPUT."<< endl;
        }
    }
}

void ALI::executeLine(){//defintion of executeLine
    if (halted) {//makes sure to check if program halted
        cout << "Program has halted." << endl;//keeps giving this till user exits
        return;
    }

    string instruction = programMemory[programCounter];//gets instruction based on program counter
    cout << "Executing instruction at line=" << programCounter << ": " << instruction << endl;//prints instruction that is gonna be executed
    executeInstruction(instruction);//executes instruction
    programCounter++;//updates values
    counter++;
    printData();//prints user info
}

void ALI::executeAll(){//defintion of executeAll that runs all functions
    int counter = 0; 
    while (programCounter < PROGRAM_MEMORY_SIZE && !halted){//runs program while we dont pass memory and isnt halted
        counter++;
        if (counter >= 1000){//when instructions go over 1000 we ask if user wants to continue
            cout << "Want to continue? Enter y or n: ";
            string input;
            cin >> input;

            if (input == "n"){//if user puts n then program halts 
                halted = true;
                break;
            } 
            else if (input == "y"){//if user puts y then program continues
                counter = 0;
            }
        }
        string instruction = programMemory[programCounter];//gets instruction
        executeInstruction(instruction);//executes instruction

        if (!halted){//while program hasnt halted then we move to next instruction 
            programCounter++;
        }
    }
    printData();//prints stats at the end
}

void ALI::executeInstruction(const string& instruction){//definition of executeInstruction which executes current instruction
    istringstream iss(instruction);
    string command;
    iss >> command;//reads in first word which is the instruction 

    if (command == "DEC"){//checks if function is declaring a variable
        string variable;
        iss >> variable;//reads in the next word which is the variable
        int address = allocateData();//moves address to next avaible one
        instructionHash[variable] = address;//stores data
        cout << "Allocated " << variable << " at address " << address << endl;
        cout << "Executed opcode: DEC" << endl;//prints to show we executed DEC
    } 
    else if (command == "LDA"){//checks if command is LDA
        string variable;
        iss >> variable;//takes in variable we want to load in
        LDA lda(this, variable);//creates lda instance
        lda.execute();//runs execute to load in variable value into accumulator
    } 
    else if (command == "LDI"){//checks if instruction is LDI
        int value;
        iss >> value;//reads in value 
        accumulator = value;//accumulator is set and value read in
    } 
    else if (command == "STR"){//checks if command is STR
        string variable;
        iss >> variable;//reads in variable to see what wants to be read in
        STR str(this);//creates instance of STR class
        str.execute(variable);//calls execute function to str value of variable into 
    } 
    else if (command == "ADD"){//checks if command is ADD
        ADD add(this);//creates instance of ADD
        add.execute();//calls exectute function that will add A and B registers and store in A
    } 
    else if (command == "SUB"){//checks if command is SUB
        SUB sub(this);//creates instance of SUB
        sub.execute();//calls exectute function that will subtract B - A registers and store in A
    } 
    else if (command == "XCH"){//checks if command is XCH 
        XCH xch(this);//creates instance of XCH command
        xch.execute();//calls execute function that swaps register values
    } 
    else if (command == "JMP"){//checks if command is JMP 
        int address;
        iss >> address;//gets line number we jump too
        JMP jmp(this, address);//creates JMP instance
        jmp.execute();//calls execute function that jmps program to line wanted
    } 
    else if (command == "JZS"){//checks if command is JZS
        int address;
        iss >> address;//read in line we will jump to
        JZS jzs(this, address);//creates an instance of JZS
        jzs.execute();//executes function that jumps if zero bit is 0
    } 
    else if (command == "HLT"){//checks if command is HLT
        HLT hlt(this);//creates HLT instance
        hlt.execute();//calls execute function that will halt the program
        halted = true;//halts the program
    } 
}

void ALI::printData() const {//definition of printData that prints relevant information of program
    cout << "Accumulator: " << accumulator << endl; //prints relevant information of registers
    cout << "Data Register: " << dataRegister << endl;
    cout << "Program Counter: " << programCounter << endl;
    cout << "Zero Bit: " << (zeroBit ? "true" : "false") << endl;

    cout << "Program Memory: ["; 
    for (int i = 0; i < PROGRAM_MEMORY_SIZE; ++i){//prints program memory which is instructions
        if (i != 0) cout << ", ";
        cout << "\"" << programMemory[i] << "\"";
    }
    cout << "]\n";

    cout << "Data Memory: [";
    for (int i = 0; i < DATA_MEMORY_SIZE; ++i){//prints variable values 
        if (i != 0) cout << ", ";
        cout << dataMemory[i];
    }
    cout << "]\n";
}

